### PR TITLE
Bug fix in FFTW, more fully support numpy FFTs and  update FFT tests to cover properly

### PR DESCRIPTION
**Pull request by spxiwh**
This patch addresses some issues found in FFT library in PyCBC. After including all FFT backends in the test suite a number of failures were noted (as described below), which have been tracked down and fixed. I also update the test code to automatically test class and function APIs for all available backends. Finally, I've added a documentation FFT page to illustrate how to use both interfaces. I have *not* yet looked at multi-threaded FFTs, which have been reported as an issue (in particular MKL multi-threaded FFTs not working).

Failures that have been fixed:

 * ~~FFTW is segfaulting on the r2c FFTs when called through the class API.~~  (Fixed after resolving bugs in how plans are being called)
 * ~~MKL gives incorrect results (I think incorrectly normalized) for the class-based API.~~ (Fixed after noticing that the class-based API does not multiply by delta_t/delta_f, whereas the function-based API does).
 * ~~numpy doesn't *have* a class-based API, and so fails.~~ (Fixed by adding a basic class-based API for numpy. This is not optimal, but is useful for people running in (for e.g.) SciServer/Colab where this is the only FFT backend easily available)
 * ~~FFTW's class-based API overwrites the input for c2r FFTs. This is currently considered a failure.~~ (While I think this still happens for large FFTs, it no longer seems to happen for the small inputs used in the test once the above things were fixed)

I'm not sure which backends the tests here will have access to.

### CODE REVIEW

**Comment by spxiwh** [comment]
@josh-willis I think this is ready for review now. I'm not sure what the appropriate level of accuracy for MKL FFTs is, but it doesn't seem to be the same as the numbers originally used in the test.

File: docs/fft.rst
Comment (ahnitz) [inline docs/fft.rst:1]: Should this be linked from the index.rst? BTW, thanks for making this! Maybe add to the set in the 'library examples' section?

Comment (spxiwh) [reply docs/fft.rst:1]: Good point. I normally just use the sidebar (and assume it will be added there automatically), but good to add this here too!

**Comment by ahnitz** [comment]
I have no other comments on this PR, but I assume you'll want sign-off from @josh-willis

File: pycbc/fft/fftw.py
Comment (josh-willis) [inline pycbc/fft/fftw.py:415]: So I don't think there's anything wrong with this change, but I also don't see that there was anything that should have been wrong with the code you're replacing, unless something was being set wrong about `fftobj.forward` (which could then be causing other problems). I've lost track of which changes were tested incrementally from others---was there a point at which the new lines 409--414 were the only change you made, and it fixed one of the problems you were seeing?  If not, it may be that some of the other changes were actually more responsible for fixing broken behavior; if so, I'm puzzled and mostly want to make sure that there isn't some remaining subtle issue that we are still missing.

When you were seeing the segfault in FFTW, could you tell whether it was in the function that created the plan, or in the function that executed that plan?

File: pycbc/fft/fftw.py
Comment (josh-willis) [inline pycbc/fft/fftw.py:388]: So this was certainly wrong before, not sure why I ever thought it was right.

As far as "why didn't we see it?", my guess is that we don't actually use the class inferface when performing R2C transforms, or C2R, since that's not in our inner loop.  In particular it does look like the C2C that is the back-bone of our matched filtering was correct, and that is almost certainly why this did not crop up.

File: docs/fft.rst
Comment (josh-willis) [inline docs/fft.rst:155]: Have you tested this?  I would only expect
```
backend_support.set_backend(['mkl'])
```
to work.  That is, `set_backend` expects a priority-ordered list.  Not that it's unreasonable to support a single entry, but I don't think the code works that way right now? I missed it if you have changed that as part of this PR.

File: pycbc/fft/fftw.py
Comment (spxiwh) [reply pycbc/fft/fftw.py:415]: This was the change that resolved the FFTW segfault. I guess that reverting this and running the tests again would prove that. I think though that refactoring this a little to remove some duplication is hiding the bug-fix in the if statement logic.

The segfault before was in R2C transforms (so C2C was good, as you highlight below). The problem was that if you have R2C you want to use the R2C plan (seen at line 423). *But* the R2C transform actually ended up at line 411 because the logic was not correct (it's a forward transform and the output is complex, so it ends up at the forward C2C plan). The logic needs to look at the type of the input as well as the output, and that's what I added. The rest is just cleanup.

The segfault actually happened when running the FFT, not when planning. The C2R in contrast did run, but gave invalid output.

File: docs/fft.rst
Comment (spxiwh) [reply docs/fft.rst:155]: Yes, you're right, this is a typo on my part. It should be what you say, I'll fix this.

**Review by josh-willis — APPROVED** [review]
Thanks. This looks good to me. I'll leave it to you to merge when ready.

**Pull request merged**
